R/cloud.R

```diff
@@ -0,0 +1,599 @@
+    elapsed <- hms::as_hms(as.integer(difftime(current_time, created_time, units = "secs")))
+
+    eta <- calc_eta(created_time, current_time, num_running, num_completed, size)
+
+    status_bar_text <- "[{num_queued}/{col_blue(num_running)}/{col_green(results$succeeded)}/{col_red(results$failed)} - {.strong {size}}] {elapsed} | ETA: {eta}"
+
+    switch(status,
+      FAILED = {
+        cli_status_clear(id = status_id, result = "failed", msg_failed = paste0("{.emph FAILED}: ", status_bar_text))
+        cli_alert("run {.fun cloud_summary} for interactive results")
+        cli_alert("run {.fun cloud_report} for markdown reports")
+        return(FALSE)
+      },
+      SUCCEEDED = {
+        cli_status_clear(id = status_id, result = "done", msg_done = paste0("{.emph SUCCEEDED}: ", status_bar_text))
+        cli_alert("run {.fun cloud_summary} for interactive results")
+        cli_alert("run {.fun cloud_report} for markdown reports")
+        return(TRUE)
+      },
+      {
+        cli::cli_status_update(id = status_id, status_bar_text)
+        return(NA)
+      }
+    )
+  }
+
+  while(is.na(res <- cloud_status_check(job_id))) {
+    Sys.sleep(update_interval)
+  }
+
+  return(invisible(res))
+}
+
+calc_eta <- function(creation_time, current_time, running, completed, total) {
+  if (completed >= total) {
+    return("Done")
+  }
+
+  infinity <- "\U221E"
+
+  if (running == 0) {
+    return(infinity)
+  }
+
+  time_diff <- as.integer(difftime(current_time, creation_time, units = "secs"))
+  to_go <- total - completed
+  secs_to_go <- time_diff / completed * to_go
+  if (secs_to_go == Inf) {
+    return(infinity)
+  }
+
+  prettyunits::pretty_sec(secs_to_go)
+}
+
+#' Fetch results from the cloud
+#'
+#' Intended mainly for internal and expert use. This function when needed by
+#' [cloud_report()] and `[cloud_summary()]`, so it is unlikely you will need to
+#' call it explicitly.
+#'
+#' @keywords internal
+#' @family cloud
+#' @inheritParams cloud_report
+#' @export
+cloud_fetch_results <- function(job_id = cloud_job(), pkg = ".") {
+  pkg <- pkg_check(pkg)
+  root <- dir_find(pkg, "root")
+
+  info <- cloud_job_describe(job_id)
+
+  job_envir <- info$jobs$container$environment[[1]]
+  aws_url <- job_envir$value[job_envir$name == "OUTPUT_S3_PATH"] %||% sprintf("s3://rstudio-revdepcheck-cloud-staging/%s/results", job_id)
+
+  out_dir <- file.path(root, job_id)
+
+  dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)
+
+  cli_alert_info("Downloading from AWS")
+  processx::run("aws", c("s3", "sync", aws_url, out_dir))
+
+  cli_alert_info("Extracting results")
+  for (file in list.files(out_dir, full.names = TRUE, pattern = ".*[.]tar[.]gz")) {
+    pkg <- sub("[.]tar[.]gz", "", basename(file))
+    if (!dir.exists(file.path(out_dir, pkg))) {
+      utils::untar(file, exdir = out_dir)
+    }
+  }
+}
+
+#' Submit a reverse dependency checking job to the cloud
+#'
+#' @param tarball A pre-built package tarball, if `NULL` a tarball will be
```

Comment:
Should this be a pair of tarballs? (so you could chose to compare against something other than the CRAN version)

Reply:
Yeah maybe in the future, but currently the service doesn't support that...